beacon/blockchain/payload_test.go

```diff
@@ -100,46 +101,36 @@ func TestOptimisticBlockBuildingRejectedBlockStateChecks(t *testing.T) {
 
 	// register async call to block building
 	var wg sync.WaitGroup          // useful to make test wait on async checks
-	var ch = make(chan struct{})   // useful to serialize build block goroutine and avoid data races
 	stateRoot := st.HashTreeRoot() // track state root before the changes done by optimistic build
 	latestHeader, err := st.GetLatestBlockHeader()
 	require.NoError(t, err)
 	latestHeader.SetStateRoot(st.HashTreeRoot())
 	expectedParentBlockRoot := latestHeader.HashTreeRoot()
 
-	b.EXPECT().RequestPayloadAsync(
-		mock.Anything, mock.Anything, mock.Anything,
-		mock.Anything, mock.Anything, mock.Anything, mock.Anything,
-	).Run(
-		func(
-			_ context.Context,
-			st *state.StateDB,
-			slot, timestamp math.U64,
-			parentBlockRoot common.Root,
-			headEth1BlockHash, finalEth1BlockHash common.ExecutionHash,
-		) {
+	b.EXPECT().RequestPayloadAsync(mock.Anything, mock.Anything).Run(
+		func(_ context.Context, r *builder.RequestPayloadData) {
 			defer wg.Done()
-			<-ch // wait for block verification to finish. This avoids data races over state reads
 			genesisHeader := genesisData.ExecutionPayloadHeader
 			genesisBlkHeader := core.GenesisBlockHeader(cs.GenesisForkVersion())
 			genesisBlkHeader.SetStateRoot(stateRoot)
 
-			require.Equal(t, timestamp, consensusTime+1)
+			require.Equal(t, r.Timestamp, consensusTime+1)
 
-			require.Equal(t, genesisHeader.GetBlockHash(), headEth1BlockHash)
+			require.Equal(t, genesisHeader.GetBlockHash(), r.HeadEth1BlockHash)
 
-			require.Equal(t, expectedParentBlockRoot, parentBlockRoot)
+			require.Equal(t, expectedParentBlockRoot, r.ParentBlockRoot)
 
-			require.Empty(t, finalEth1BlockHash)          // this is first block post genesis
-			require.Equal(t, constants.GenesisSlot, slot) // genesis slot in state
-			var stateSlot math.Slot
-			stateSlot, err = st.GetSlot()
-			require.NoError(t, err)
-			require.Equal(t, constants.GenesisSlot, stateSlot)
+			require.Empty(t, r.FinalEth1BlockHash)          // this is first block post genesis
+			require.Equal(t, constants.GenesisSlot, r.Slot) // genesis slot in state
 		},
 	).Return(nil, common.Version{0xff}, errors.New("does not matter")) // return values do not really matter in this test
 	wg.Add(1)
 
+	// check slot pre test
+	slot, err := st.GetSlot()
+	require.NoError(t, err)
+	require.Equal(t, constants.GenesisSlot, slot)
```

Comment:
these checks are not in `main` because there we don't cache state and we don't really care about state pollution.
In preparation of state caching we are stricter in the kind of invariants that optimistic block building should provide